cf-oscore/src/test/java/org/eclipse/californium/oscore/OSCoreObserveTest.java

```diff
@@ -0,0 +1,340 @@
+/*******************************************************************************
+ * Copyright (c) 2019 RISE SICS and others.
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ *
+ * This test class is based on org.eclipse.californium.core.test.SmallServerClientTest
+ * 
+ * Contributors: 
+ *    Rikard Höglund (RISE SICS) - testing OSCORE Observe messages
+ ******************************************************************************/
+package org.eclipse.californium.oscore;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.util.Random;
+import java.util.Timer;
+import java.util.TimerTask;
+
+import org.eclipse.californium.category.Large;
+import org.eclipse.californium.core.CoapClient;
+import org.eclipse.californium.core.CoapHandler;
+import org.eclipse.californium.core.CoapObserveRelation;
+import org.eclipse.californium.core.CoapResource;
+import org.eclipse.californium.core.CoapResponse;
+import org.eclipse.californium.core.CoapServer;
+import org.eclipse.californium.core.coap.Request;
+import org.eclipse.californium.core.coap.Response;
+import org.eclipse.californium.core.coap.CoAP.Code;
+import org.eclipse.californium.core.coap.CoAP.ResponseCode;
+import org.eclipse.californium.core.coap.CoAP.Type;
+import org.eclipse.californium.core.coap.MediaTypeRegistry;
+import org.eclipse.californium.core.network.CoapEndpoint;
+import org.eclipse.californium.core.network.EndpointManager;
+import org.eclipse.californium.core.server.resources.CoapExchange;
+import org.eclipse.californium.cose.AlgorithmID;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+
+/**
+ * Performs tests of Observe message exchanges between an OSCORE server and client
+ * Based on interoperability test specification created by Ericsson:
+ * https://ericssonresearch.github.io/OSCOAP/test-spec5.html
+ * 
+ */
+@Category(Large.class)
+public class OSCoreObserveTest {
+	private CoapServer server;
```

Comment:
> java.io.IOError: java.net.SocketException: Use org.eclipse.californium.elements.rule.NetworkRule to define DatagramSocket behaviour!

Last fix for the above error to run `mvn test` successfully for the complete californium project: 

```
public class OSCoreObserveTest {
   @ClassRule
   public static CoapNetworkRule network = new CoapNetworkRule(CoapNetworkRule.Mode.DIRECT, CoapNetworkRule.Mode.NATIVE);
```

That's required, if a datagram socket is involved in the tests. It enables the tests to replace the datagram socket implementation by a in memory variant, which ensures, that records are not dropped unintended. The rule enables the test execution to also select tests, which can only be execute with the "native" datagram socket implementation or the "direct". The rule also ensures, that changes in `NetworkConfig`are rolled back on the end of the test.

Reply:
Alright, thanks for the information. I had missed that it was needed. It's now been added via commit 7c901cea597079c0bb47004bcb8a4805c7b5c2e5